File: Data Structures/Hash/LV2_42578.cpp
#include <string>
#include <vector>
#include <iostream>
#include <map>
#include <set>

// 7:25 ~ 7:34

using namespace std;

int solution(vector<vector<string>> clothes) {
    int answer = 1;
    
    // (의상의 종류 - 의상)을 저장할 map
    map <string, set<string>> closet;
    
    // 의상을 종류별로 map의 set에 저장
    for (auto costume : clothes) {
        closet[costume[1]].insert(costume[0]);
    }
    
    // 서로 다른 의상의 경우의 수 구하기
    for (auto dress : closet) {
        answer *= (dress.second.size() + 1);
    }
    
    answer--;
    
    return answer;
}

File: Data Structures/Hash/LV2_42578.py
def solution(clothes):
    # 해시 풀이
    answer = 1
    clothes_list = {}
    for c, t in clothes:
        if t not in clothes_list:
            clothes_list[t] = 2
        else:
            clothes_list[t] += 1
    for p in clothes_list.values():
        answer *= p
    return answer - 1
    
    # pythonic 풀이
    from collections import Counter
    from functools import reduce
    clothes_count = Counter([kind for cloth, kind in clothes])
    answer = reduce(lambda x, y : x * (y+1), clothes_count.values(), 1) - 1
    return answer
